doc/reference/kernel/timing/clocks.rst

```diff
@@ -1,161 +1,242 @@
+  missed.  Note that the timeout value passed here is in a delta from
+  current time, but that does not absolve the driver of the
+  requirement to provide ticks at a steady rate over time.  Naive
+  implementations of this function are subject to bugs where the
+  fractional tick gets "reset" incorrectly and causes clock skew.
+
+* The driver is expected to provide a ``z_clock_elapsed()`` call which
+  provides a current indication of how many ticks have elapsed (as
+  compared to a real world clock) since the last call to
+  ``z_clock_announce()``, which the kernel needs to test newly
+  arriving timeouts for expiration.
+
+Note that a natural implementation of this API results in a "tickless"
+kernel, which receives and processes timer interrupts only for
+registered events, relying on programmable hardware counters to
+provide irregular interrupts.  But a traditional, "ticked" or "dumb"
+counter driver can be trivially implemented also:
+
+* The driver can receive interrupts at a regular rate corresponding to
+  the OS tick rate, calling z_clock_anounce() with an argument of one
+  each time.
+
+* The driver can ignore calls to ``z_clock_set_timeout()``, as every
+  tick will be announced regardless of timeout status.
+
+* The driver can return zero for every call to ``z_clock_elapsed()``
+  as no more than one tick can be detected as having elapsed (because
+  otherwise an interrupt would have been received).
+
+SMP Details
+-----------
+
+In general, the timer API described above does not change when run in
+a multiprocessor context.  The kernel will internally synchronize all
+access appropriately, and ensure that all critical sections are small
+and minimal.  But some notes are important to detail:
+
+* Zephyr is agnostic about which CPU services timer interrupts.  It is
+  not illegal (though probably undesirable in some circumstances) to
+  have every timer interrupt handled on a single processor.  Existing
+  SMP architectures implement symmetric timer drivers.
+
+* The ``z_clock_announce()`` call is expected to be be globally
+  synchronized at the driver level.  The kernel does not do any
+  per-CPU tracking, and expects that if two timer interrupts fire near
+  simultaneously, that only one will provide the current tick count to
+  the timing subsystem.  The other may legally provide a tick count of
+  zero if no ticks have elapsed.  It should not "skip" the announce
+  call because of timeslicing requirements (see below).
+
+* Some SMP hardware uses a single, global timer device, others use a
+  per-CPU counter.  The complexity here (for example: ensuring counter
+  synchronization between CPUs) is expected to be managed by the
+  driver, not the kernel.
+
+* The next timeout value passed back to the driver via
+  ``z_clock_set_timeout()`` is done identically for every CPU.  So by
+  default, every CPU will see simultaneous timer interrupts for every
+  event, even though by definition only one of them.  This is probably
```

Comment:
This sentence seems to be cut mid way (verb missing): ", even though by definition only one of them"